--- csrc/cpp_itfs/mla/asm_mla_decode_fwd.py
+++ csrc/cpp_itfs/mla/asm_mla_decode_fwd.py
@@ -3,12 +3,14 @@ from csrc.cpp_itfs.utils import compile_template_op, transfer_hsaco, AITER_CORE_
 from aiter.aot.triton.compile import compile_kernel
 import triton
 
+import os
 
 MD_NAME = "asm_mla_decode_fwd"
 warpSize = 64
 with open(f"{AITER_CORE_DIR}/csrc/cpp_itfs/mla/asm_mla_decode_fwd.cpp.jinja", "r") as f:
     src_template = Template(f.read())
 
+os.chdir(os.path.dirname(__file__))
 
 def compile(hsaco_path: str, page_size: int, q_dtype: str, kv_dtype: str, num_kv_splits:int, v_head_dim:int, func_name: str = None):
     if func_name is None: