diff --git a/aiter/ops/mha.py b/aiter/ops/mha.py
index 0892939..b84e620 100644
--- a/aiter/ops/mha.py
+++ b/aiter/ops/mha.py
@@ -1014,7 +1014,7 @@ def _flash_attn_forward(
         ret = ret and (not swa)
         ret = ret and (q.dtype == dtypes.bf16)
         ret = ret and ((return_lse and gfx == "gfx950") or (gfx == "gfx942"))
-        return ret
+        return
 
     q, k, v = [maybe_contiguous(x) for x in (q, k, v)]
     if can_impl_fmha_v3_fwd():