--- aiter/jit/optCompilerConfig.json
+++ aiter/jit/optCompilerConfig.json
@@ -619,6 +619,7 @@
         "verbose": "False",
         "hip_clang_path": "os.environ.get('MHA_HIP_CLANG_PATH')",
         "blob_gen_cmd": [
+            "f'{get_asm_dir()}/fmha_v3_fwd/codegen.py --output_dir {{}}'",
             "f'{CK_DIR}/example/ck_tile/01_fmha/generate.py -d fwd --receipt 600 --output_dir {{}}'",
             "f'{AITER_CSRC_DIR}/cpp_itfs/mha_fwd_generate.py --receipt 3 --output_dir {{}}'"
         ]

--- csrc/cpp_itfs/mha_fwd_generate.py
+++ csrc/cpp_itfs/mha_fwd_generate.py
@@ -150,7 +150,7 @@ COMBINED_API = """t = fmha_fwd_v3(traits, args, stream_config);
 API_MAP = {
     1: FMHA_FWD_API.format(F_inner_dispatch=V3_API),
     2: FMHA_FWD_API.format(F_inner_dispatch=V2_API),
-    3: FMHA_FWD_API.format(F_inner_dispatch=V2_API) + FMHA_FWD_SPLITKV_API,
+    3: FMHA_FWD_API.format(F_inner_dispatch=COMBINED_API),
     4: FMHA_BATCH_PREFILL_API,
     5: FMHA_FWD_API.format(F_inner_dispatch=COMBINED_API)
     + FMHA_FWD_SPLITKV_API