--- csrc/py_itfs_cu/asm_pa.cu
+++ csrc/py_itfs_cu/asm_pa.cu
@@ -97,7 +97,7 @@ torch::Tensor pa_fwd(torch::Tensor& Q, //   [num_seqs, num_heads, head_size]
     int num_heads       = Q.size(1);
     int head_size       = Q.size(2);
     int num_kv_heads    = K.size(1);
-    int block_size      = K.size(3);
+    int block_size      = K.size(2);
     const int gqa_ratio = num_heads / num_kv_heads;
     TORCH_CHECK(block_size == 16, __func__, " for now only support block_size == 16");

