--- aiter/jit/core.py
+++ aiter/jit/core.py
@@ -114,10 +114,12 @@ def get_user_jit_dir():
         shutil.copytree(this_dir, home_jit_dir)
     return home_jit_dir
 
+CK_DIR = os.environ.get("CK_DIR", f"{AITER_ROOT_DIR}/3rdparty/composable_kernel")
 
 bd_dir = f"{get_user_jit_dir()}/build"
 # copy ck to build, thus hippify under bd_dir
 if multiprocessing.current_process().name == "MainProcess":
+    shutil.copytree(CK_DIR, f'{bd_dir}/ck', dirs_exist_ok=True)
     os.makedirs(bd_dir, exist_ok=True)
     # if os.path.exists(f"{bd_dir}/ck/library"):
     #     shutil.rmtree(f"{bd_dir}/ck/library")
