diff --git python/pip_install/extract_wheels/lib/bazel.py python/pip_install/extract_wheels/lib/bazel.py
index ef0d6e8..10c193d 100644
--- python/pip_install/extract_wheels/lib/bazel.py
+++ python/pip_install/extract_wheels/lib/bazel.py
@@ -183,6 +183,10 @@ def extract_wheel(
     sanitised_wheel_file_dependencies = [
         '"//%s:%s"' % (sanitise_name(d), WHEEL_FILE_LABEL) for d in whl_deps
     ]
+    
+    if whl.name == "triton":
+        sanitised_dependencies = [s for s in sanitised_dependencies if 'torch' not in s]
+        sanitised_wheel_file_dependencies = [s for s in sanitised_wheel_file_dependencies if 'torch' not in s]
 
     with open(os.path.join(directory, "BUILD"), "w") as build_file:
         contents = generate_build_file_contents(
diff --git python/pip_install/repositories.bzl python/pip_install/repositories.bzl
index 828ab53..dc216c9 100644
--- python/pip_install/repositories.bzl
+++ python/pip_install/repositories.bzl
@@ -16,13 +16,13 @@ _RULE_DEPS = [
     ),
     (
         "pypi__setuptools",
-        "https://files.pythonhosted.org/packages/ab/b5/3679d7c98be5b65fa5522671ef437b792d909cf3908ba54fe9eca5d2a766/setuptools-44.1.0-py2.py3-none-any.whl",
-        "992728077ca19db6598072414fb83e0a284aca1253aaf2e24bb1e55ee6db1a30",
+        "https://files.pythonhosted.org/packages/55/3a/5121b58b578a598b269537e09a316ad2a94fdd561a2c6eb75cd68578cc6b/setuptools-69.0.3-py3-none-any.whl",
+        "385eb4edd9c9d5c17540511303e39a147ce2fc04bc55289c322b9e5904fe2c05",
     ),
     (
         "pypi__wheel",
-        "https://files.pythonhosted.org/packages/c9/0b/e0fd299d93cd9331657f415085a4956422959897b333e3791dde40bd711d/wheel-0.36.1-py2.py3-none-any.whl",
-        "906864fb722c0ab5f2f9c35b2c65e3af3c009402c108a709c0aca27bc2c9187b",
+        "https://files.pythonhosted.org/packages/bd/7c/d38a0b30ce22fc26ed7dbc087c6d00851fb3395e9d0dac40bec1f905030c/wheel-0.38.4-py3-none-any.whl",
+        "b60533f3f5d530e971d6737ca6d58681ee434818fab630c83a734bb10c083ce8",
     ),
 ]
 
diff --git python/requirements.txt python/requirements.txt
index 164fa26..743bc87 100644
--- python/requirements.txt
+++ python/requirements.txt
@@ -1,6 +1,6 @@
 pip==9.0.3
-setuptools==44.0.0
-wheel==0.30.0a0
+setuptools==69.0.3
+wheel==0.38.4
 
 # For tests
 mock==2.0.0
