diff --git python/pip_install/extract_wheels/__init__.py python/pip_install/extract_wheels/__init__.py
index fe8b8ef..a64f7f8 100644
--- python/pip_install/extract_wheels/__init__.py
+++ python/pip_install/extract_wheels/__init__.py
@@ -79,14 +79,17 @@ def main() -> None:
     )
     args = parser.parse_args()

-    pip_args = [sys.executable, "-m", "pip", "--isolated", "wheel", "-r", args.requirements]
+    requirement_files = args.requirements.split(",")
+    pip_args = [sys.executable, "-m", "pip", "--isolated", "wheel"]
+    for requirement_file in requirement_files:
+        pip_args.extend(['-r', requirement_file])
     if args.extra_pip_args:
         pip_args += json.loads(args.extra_pip_args)["args"]

     # Assumes any errors are logged by pip so do nothing. This command will fail if pip fails
     subprocess.run(pip_args, check=True)

-    extras = requirements.parse_extras(args.requirements)
+    extras = requirements.parse_extras(requirement_files)

     if args.pip_data_exclude:
         pip_data_exclude = json.loads(args.pip_data_exclude)["exclude"]
diff --git python/pip_install/extract_wheels/lib/bazel.py python/pip_install/extract_wheels/lib/bazel.py
index ef0d6e8..10c193d 100644
--- python/pip_install/extract_wheels/lib/bazel.py
+++ python/pip_install/extract_wheels/lib/bazel.py
@@ -183,6 +183,10 @@ def extract_wheel(
     sanitised_wheel_file_dependencies = [
         '"//%s:%s"' % (sanitise_name(d), WHEEL_FILE_LABEL) for d in whl_deps
     ]
+
+    if whl.name == "triton":
+        sanitised_dependencies = [s for s in sanitised_dependencies if 'torch' not in s]
+        sanitised_wheel_file_dependencies = [s for s in sanitised_wheel_file_dependencies if 'torch' not in s]

     with open(os.path.join(directory, "BUILD"), "w") as build_file:
         contents = generate_build_file_contents(
diff --git python/pip_install/extract_wheels/lib/requirements.py python/pip_install/extract_wheels/lib/requirements.py
index e246379..583796c 100644
--- python/pip_install/extract_wheels/lib/requirements.py
+++ python/pip_install/extract_wheels/lib/requirements.py
@@ -1,8 +1,8 @@
 import re
-from typing import Dict, Set, Tuple, Optional
+from typing import Dict, Set, Tuple, Optional, List


-def parse_extras(requirements_path: str) -> Dict[str, Set[str]]:
+def parse_extras(requirements_path: List[str]) -> Dict[str, Set[str]]:
     """Parse over the requirements.txt file to find extras requested.

     Args:
@@ -13,12 +13,13 @@ def parse_extras(requirements_path: str) -> Dict[str, Set[str]]:
     """

     extras_requested = {}
-    with open(requirements_path, "r") as requirements:
-        # Merge all backslash line continuations so we parse each requirement as a single line.
-        for line in requirements.read().replace("\\\n", "").split("\n"):
-            requirement, extras = _parse_requirement_for_extra(line)
-            if requirement and extras:
-                extras_requested[requirement] = extras
+    for requirement_path in requirements_path:
+        with open(requirement_path, "r") as requirements:
+            # Merge all backslash line continuations so we parse each requirement as a single line.
+            for line in requirements.read().replace("\\\n", "").split("\n"):
+                requirement, extras = _parse_requirement_for_extra(line)
+                if requirement and extras:
+                    extras_requested[requirement] = extras

     return extras_requested

diff --git python/pip_install/pip_repository.bzl python/pip_install/pip_repository.bzl
index b2cb0a8..5da37b9 100644
--- python/pip_install/pip_repository.bzl
+++ python/pip_install/pip_repository.bzl
@@ -25,12 +25,14 @@ def _pip_repository_impl(rctx):
     separator = ":" if not "windows" in rctx.os.name.lower() else ";"
     pypath = separator.join([str(p) for p in [rules_root] + thirdparty_roots])

+    requirements_files = [str(rctx.path(requirement)) for requirement in rctx.attr.requirements]
+
     args = [
         python_interpreter,
         "-m",
         "python.pip_install.extract_wheels",
         "--requirements",
-        rctx.path(rctx.attr.requirements),
+        ",".join(requirements_files),
         "--repo",
         "@%s" % rctx.attr.name,
     ]
@@ -93,8 +95,8 @@ python_interpreter.
             default = True,
             doc = "If True, suppress printing stdout and stderr output to the terminal.",
         ),
-        "requirements": attr.label(
-            allow_single_file = True,
+        "requirements": attr.label_list(
+            allow_files = ["*.txt"],
             mandatory = True,
             doc = "A 'requirements.txt' pip requirements file.",
         ),
diff --git python/pip_install/repositories.bzl python/pip_install/repositories.bzl
index 828ab53..2067e1e 100644
--- python/pip_install/repositories.bzl
+++ python/pip_install/repositories.bzl
@@ -11,18 +11,18 @@ _RULE_DEPS = [
     ),
     (
         "pypi__pkginfo",
-        "https://files.pythonhosted.org/packages/4f/3c/535287349af1b117e082f8e77feca52fbe2fdf61ef1e6da6bcc2a72a3a79/pkginfo-1.6.1-py2.py3-none-any.whl",
-        "ce14d7296c673dc4c61c759a0b6c14bae34e34eb819c0017bb6ca5b7292c56e9",
+        "https://files.pythonhosted.org/packages/56/09/054aea9b7534a15ad38a363a2bd974c20646ab1582a387a95b8df1bfea1c/pkginfo-1.10.0-py3-none-any.whl",
+        "889a6da2ed7ffc58ab5b900d888ddce90bce912f2d2de1dc1c26f4cb9fe65097",
     ),
     (
         "pypi__setuptools",
-        "https://files.pythonhosted.org/packages/ab/b5/3679d7c98be5b65fa5522671ef437b792d909cf3908ba54fe9eca5d2a766/setuptools-44.1.0-py2.py3-none-any.whl",
-        "992728077ca19db6598072414fb83e0a284aca1253aaf2e24bb1e55ee6db1a30",
+        "https://files.pythonhosted.org/packages/55/3a/5121b58b578a598b269537e09a316ad2a94fdd561a2c6eb75cd68578cc6b/setuptools-69.0.3-py3-none-any.whl",
+        "385eb4edd9c9d5c17540511303e39a147ce2fc04bc55289c322b9e5904fe2c05",
     ),
     (
         "pypi__wheel",
-        "https://files.pythonhosted.org/packages/c9/0b/e0fd299d93cd9331657f415085a4956422959897b333e3791dde40bd711d/wheel-0.36.1-py2.py3-none-any.whl",
-        "906864fb722c0ab5f2f9c35b2c65e3af3c009402c108a709c0aca27bc2c9187b",
+        "https://files.pythonhosted.org/packages/bd/7c/d38a0b30ce22fc26ed7dbc087c6d00851fb3395e9d0dac40bec1f905030c/wheel-0.38.4-py3-none-any.whl",
+        "b60533f3f5d530e971d6737ca6d58681ee434818fab630c83a734bb10c083ce8",
     ),
 ]

diff --git python/requirements.txt python/requirements.txt
index 164fa26..743bc87 100644
--- python/requirements.txt
+++ python/requirements.txt
@@ -1,6 +1,6 @@
 pip==9.0.3
-setuptools==44.0.0
-wheel==0.30.0a0
+setuptools==69.0.3
+wheel==0.38.4

 # For tests
 mock==2.0.0
