commit 72b94b67c2b57bb98df191db542f1b6739ebbac9
Author: liukan.lk <liukan.lk@alibaba-inc.com>
Date:   Fri Dec 27 16:29:47 2024 +0800

    fix compile

diff --git include/flashinfer/gemm/bmm_fp8.cuh include/flashinfer/gemm/bmm_fp8.cuh
index 7778934..50ce7f9 100644
--- include/flashinfer/gemm/bmm_fp8.cuh
+++ include/flashinfer/gemm/bmm_fp8.cuh
@@ -45,7 +45,7 @@
 #else
 #define FLASHINFER_CUBLAS_CALL(EXPR)  \
   {                                   \
-    cublasStatus_t e = (EXPR);        \
+    auto e = (EXPR);           \
     if (e != CUBLAS_STATUS_SUCCESS) { \
       return e;                       \
     }                                 \
