From b99a7c590f0c600099ea2474692b8d4fcb9a8bf4 Mon Sep 17 00:00:00 2001
From: "liukan.lk" <liukan.lk@alibaba-inc.com>
Date: Thu, 16 Jan 2025 15:13:39 +0800
Subject: [PATCH 2/2] dispatch group size


diff --git a/include/flashinfer/utils.cuh b/include/flashinfer/utils.cuh
index b36145f..73b548e 100644
--- include/flashinfer/utils.cuh
+++ include/flashinfer/utils.cuh
@@ -132,6 +132,24 @@
   } else if (group_size == 4) {                              \
     constexpr size_t GROUP_SIZE = 4;                         \
     __VA_ARGS__                                              \
+  } else if (group_size == 5) {                              \
+    constexpr size_t GROUP_SIZE = 5;                         \
+    __VA_ARGS__                                              \
+  } else if (group_size == 6) {                              \
+    constexpr size_t GROUP_SIZE = 6;                         \
+    __VA_ARGS__                                              \
+  } else if (group_size == 7) {                              \
+    constexpr size_t GROUP_SIZE = 7;                         \
+    __VA_ARGS__                                              \
+  } else if (group_size == 9) {                              \
+    constexpr size_t GROUP_SIZE = 9;                         \
+    __VA_ARGS__                                              \
+  } else if (group_size == 10) {                             \
+    constexpr size_t GROUP_SIZE = 10;                        \
+    __VA_ARGS__                                              \
+  } else if (group_size == 16) {                             \
+    constexpr size_t GROUP_SIZE = 16;                        \
+    __VA_ARGS__                                              \
   } else if (group_size == 8) {                              \
     constexpr size_t GROUP_SIZE = 8;                         \
     __VA_ARGS__                                              \
-- 
2.19.1.6.gb485710b

