diff --git aot_build_utils/generate_dispatch_inc.py aot_build_utils/generate_dispatch_inc.py
index 3f7ad94..f4f340b 100644
--- aot_build_utils/generate_dispatch_inc.py
+++ aot_build_utils/generate_dispatch_inc.py
@@ -99,6 +99,9 @@ if __name__ == "__main__":
     parser.add_argument(
         "--path", type=str, required=True, help="Path to the dispatch inc file"
     )
+    parser.add_argument(
+        "--head_dims_sm90", type=str, required=True, nargs="+", help="Head dimensions in format of 'head_dim_qk,head_dim_vo'",
+    )
     parser.add_argument(
         "--head_dims", type=int, required=True, nargs="+", help="Head dimensions"
     )
@@ -124,6 +127,7 @@ if __name__ == "__main__":
         help="Mask modes",
     )
     args = parser.parse_args()
+    args.head_dims_sm90 = [tuple(map(int, x.split(","))) for x in args.head_dims_sm90]
     print(args)
     with open(Path(args.path), "w") as f:
         f.write(get_dispatch_inc_str(args))
diff --git include/flashinfer/attention/cascade.cuh include/flashinfer/attention/cascade.cuh
index 96b769f..766b6e2 100644
--- include/flashinfer/attention/cascade.cuh
+++ include/flashinfer/attention/cascade.cuh
@@ -25,6 +25,9 @@
 
 namespace flashinfer {
 
+DEFINE_HAS_MEMBER(maybe_q_rope_offset)
+DEFINE_HAS_MEMBER(maybe_k_rope_offset)
+
 namespace cg = cooperative_groups;
 using cp_async::PrefetchMode;
 using cp_async::SharedMemFillMode;
diff --git include/flashinfer/attention/decode.cuh include/flashinfer/attention/decode.cuh
index f2c7a09..d8dc1a5 100644
--- include/flashinfer/attention/decode.cuh
+++ include/flashinfer/attention/decode.cuh
@@ -33,8 +33,6 @@
 
 namespace flashinfer {
 
-DEFINE_HAS_MEMBER(maybe_q_rope_offset)
-
 namespace cg = cooperative_groups;
 using cp_async::PrefetchMode;
 using cp_async::SharedMemFillMode;
diff --git include/flashinfer/attention/prefill.cuh include/flashinfer/attention/prefill.cuh
index bb82507..800d3fd 100644
--- include/flashinfer/attention/prefill.cuh
+++ include/flashinfer/attention/prefill.cuh
@@ -36,9 +36,6 @@
 
 namespace flashinfer {
 
-DEFINE_HAS_MEMBER(maybe_q_rope_offset)
-DEFINE_HAS_MEMBER(maybe_k_rope_offset)
-
 namespace cg = cooperative_groups;
 using cp_async::SharedMemFillMode;
 using mma::MMAMode;
