--- example/ck_tile/01_fmha/codegen/cpp_symbol_map.py	2024-07-15 05:21:03.583380865 +0000
+++ example/ck_tile/01_fmha/codegen/cpp_symbol_map.py	2024-07-15 05:11:37.089034571 +0000
@@ -4,8 +4,8 @@
 
 DTYPE_MAP = {
     "fp16": "ck_tile::fp16_t",
-    "bf16": "ck_tile::bf16_t",
-    "fp8" : "ck_tile::fp8_t"
+    "bf16": "ck_tile::bf16_t", # enable bf16
+    # "fp8" : "ck_tile::fp8_t"
 }
 
 MASK_IMPL = {
@@ -55,13 +55,13 @@
 
 BIAS_MAP = {
     "no" : "ck_tile::BlockAttentionBiasEnum::NO_BIAS",
-    "bias"  : "ck_tile::BlockAttentionBiasEnum::ELEMENTWISE_BIAS",
-    "alibi" : "ck_tile::BlockAttentionBiasEnum::ALIBI"
+    # "bias"  : "ck_tile::BlockAttentionBiasEnum::ELEMENTWISE_BIAS",
+     "alibi" : "ck_tile::BlockAttentionBiasEnum::ALIBI"
 }
 
 # TODO: this is ugly
 BIAS_CHECK_MAP = {
     "no" : "bias_enum::no_bias",
-    "bias"  : "bias_enum::elementwise_bias",
-    "alibi" : "bias_enum::alibi"
+    # "bias"  : "bias_enum::elementwise_bias",
+     "alibi" : "bias_enum::alibi"
 }
@@ -84,5 +84,5 @@

 MODE_MAP = {
-    "batch" : "false",
+    # "batch" : "false",
     "group" : "true"
 }