--- example/ck_tile/10_rmsnorm2d/CMakeLists.txt
+++ example/ck_tile/10_rmsnorm2d/CMakeLists.txt
@@ -25,10 +25,12 @@
 
 set(TILE_RMSNORM2D_FWD "tile_rmsnorm2d_fwd")
 
-message("adding ${TILE_RMSNORM2D_FWD}")
-add_executable(${TILE_RMSNORM2D_FWD} EXCLUDE_FROM_ALL rmsnorm2d_fwd.cpp)
+# message("adding ${TILE_RMSNORM2D_FWD}")
+# add_executable(${TILE_RMSNORM2D_FWD} EXCLUDE_FROM_ALL rmsnorm2d_fwd.cpp)
+message("adding shared library ${TILE_RMSNORM2D_FWD}")
+add_library(${TILE_RMSNORM2D_FWD} SHARED ${RMSNORM2D_FWD_GEN_BLOBS})
 target_include_directories(${TILE_RMSNORM2D_FWD} PRIVATE ${CMAKE_CURRENT_LIST_DIR})
-target_sources(${TILE_RMSNORM2D_FWD} PRIVATE ${RMSNORM2D_FWD_GEN_BLOBS})
+# target_sources(${TILE_RMSNORM2D_FWD} PRIVATE ${RMSNORM2D_FWD_GEN_BLOBS})
 
 set(TILE_RMSNORM2D_FWD_COMPILE_OPTIONS)
 
