diff --git aios/network/anet/BUILD aios/network/anet/BUILD
index 3774514c..78d172c5 100644
--- aios/network/anet/BUILD
+++ aios/network/anet/BUILD
@@ -5,6 +5,7 @@ cc_library(
     srcs=glob(['*.cpp']),
     hdrs=glob(['*.h']),
     deps=['//aios/alog:alog', '//aios/autil:env_util'],
-    copts=['-Wformat-truncation=0'],
+    copts=['-Wformat-truncation=0', '-mcpu=native'],
+    linkopts=['-ldl'],
     alwayslink=True
 )