--- a/mooncake-transfer-engine/src/config.cpp
+++ b/mooncake-transfer-engine/src/config.cpp
@@ -36,7 +36,7 @@
         std::getenv("MC_NUM_COMP_CHANNELS_PER_CTX");
     if (num_comp_channels_per_ctx_env) {
         int val = atoi(num_comp_channels_per_ctx_env);
-        if (val > 0 && val < 256)
+        if (val >= 0 && val < 256)
             config.num_comp_channels_per_ctx = val;
         else
             LOG(WARNING) << "Ignore value from environment variable "
