diff -ruN a/mooncake-transfer-engine/include/transport/rdma_transport/rdma_context.h b/mooncake-transfer-engine/include/transport/rdma_transport/rdma_context.h
--- a/mooncake-transfer-engine/include/transport/rdma_transport/rdma_context.h	2026-04-25 00:28:38.600925542 +0800
+++ b/mooncake-transfer-engine/include/transport/rdma_transport/rdma_context.h	2026-04-25 00:28:38.626925901 +0800
@@ -169,6 +169,7 @@
     RdmaTransport &engine_;
 
     ibv_context *context_ = nullptr;
+    ibv_device **device_list_ = nullptr;
     ibv_pd *pd_ = nullptr;
     uint64_t max_mr_size;
     int event_fd_ = -1;
diff -ruN a/mooncake-transfer-engine/src/transport/rdma_transport/rdma_context.cpp b/mooncake-transfer-engine/src/transport/rdma_transport/rdma_context.cpp
--- a/mooncake-transfer-engine/src/transport/rdma_transport/rdma_context.cpp	2026-04-25 00:28:38.581925280 +0800
+++ b/mooncake-transfer-engine/src/transport/rdma_transport/rdma_context.cpp	2026-04-25 00:28:38.626925901 +0800
@@ -214,6 +214,10 @@
             PLOG(ERROR) << "Failed to close device context";
         context_ = nullptr;
     }
+    if (device_list_) {
+        ibv_free_device_list(device_list_);
+        device_list_ = nullptr;
+    }
 
     return 0;
 }
@@ -641,7 +645,7 @@
         active_speed_ = attr.active_speed;
         gid_index_ = gid_index;
 
-        ibv_free_device_list(devices);
+        device_list_ = devices;
         return 0;
 
     cleanup_context_and_devices:
