--- a/csrc/fast_hadamard_transform.cpp
+++ b/csrc/fast_hadamard_transform.cpp
@@ -239,9 +239,11 @@
     return out.reshape(shapes_og);
 }
 
+#ifndef RTP_LLM_EMBEDDED_FHT
 PYBIND11_MODULE(TORCH_EXTENSION_NAME, m) {
     m.def("fast_hadamard_transform", &fast_hadamard_transform, "Fast Hadamard transform");
     m.def("fast_hadamard_transform_12N", &fast_hadamard_transform_20N, "Fast Hadamard transform with dimension = 12 * power of 2");
     m.def("fast_hadamard_transform_20N", &fast_hadamard_transform_20N, "Fast Hadamard transform with dimension = 20 * power of 2");
     m.def("fast_hadamard_transform_28N", &fast_hadamard_transform_28N, "Fast Hadamard transform with dimension = 28 * power of 2");
 }
+#endif
